impeller/geometry/matrix_unittests.cc

```diff
@@ -160,5 +161,20 @@ TEST(MatrixTest, TransformHomogenous) {
             Vector3(32.0f, 33.0f, 41.0f));
 }
 
+// Verifies a translate scale matrix doesn't need to compute sqrt(pow(scale, 2))
+TEST(MatrixTest, GetMaxBasisXYWithLargeAndSmallScalingFactor) {
+  Matrix m = Matrix::MakeScale({2.625e+20, 2.625e+20, 1});
+  EXPECT_NEAR(m.GetMaxBasisLengthXY(), 2.625e+20, 1e+20);
+
+  m = Matrix::MakeScale({2.625e-20, 2.625e-20, 1});
+  EXPECT_NEAR(m.GetMaxBasisLengthXY(), 2.625e-20, 1e-20);
+}
+
+TEST(MatrixTest, GetMaxBasisXYWithLargeAndSmallScalingFactorNonScaleTranslate) {
+  Matrix m = Matrix::MakeScale({2.625e+20, 2.625e+20, 1}) *
+             Matrix::MakeRotationX(Radians(kPi / 2));
+  EXPECT_TRUE(std::isinf(m.GetMaxBasisLengthXY()));
```

Comment:
This is failing, perhaps that isn't enough to create infinities on some platforms?

Could this be a case of a platform treating results that are too big for single precision as large, but not infinite values? I looked for answers and discovered that a compliant implementation is allowed to round to `::max` if it is close enough. I think they said if it is closer to `::max` than `::nextafter(::max)`. Perhaps use a larger exponent here to be sure (one that doesn't overflow `::max` but squaring it overflows by a larger amount?

Reply:
oh, I think its just not tripping on the scale/translate check anymore. I should just set one of matrix entires that will trigger it directly.

Reply:
Is the math done in doubles and so the sqrt reduces it back into 32-bit float before the return?

Reply:
oh, so the matrix is something like:

```
[ERROR:flutter/impeller/geometry/matrix_unittests.cc(176)] (
       2.000000,       0.000000,       0.000000,       0.000000,
       0.000000,       0.000000,      -2.000000,       0.000000,
       0.000000,       1.000000,       0.000000,       0.000000,
       0.000000,       0.000000,       0.000000,       1.000000,
)
```

am I making the right check here?

```
    if (e[0][1] == 0 && e[1][0] == 0) {
      return std::max(m[0], m[5]);
    }
    ```

Reply:
You rotated around the X axis which moved it into the 3rd dimension. We're seeing it end-on ad so the Y scale went to zero as everything is in the Y=0 plane.

Reply:
Another way to look at this is that if the optimization doesn't trigger then only these 4 entries will be used to compute the value. So, if 2 of them are 0 (2 that happen to be 0 a lot), then you can simplify the calculation.

One question is why you are mixing `e` and `m` representations? `e[0][0]` should be optimized by the compiler to "base of union + 0" and `e[1][1]` should be optimized to "base of union + 5" so `m[0]` and `m[5]` aren't any more efficient, are they?

Reply:
No Its not an efficiency thing, I was just being a bit lazy. So otherwise this code is correct then?

Reply:
"optimization" doesn't necessarily mean an efficiency thing. You did it to do less work because more work was causing an overflow. It's still a more direct path to the answer.

The new test code is one way to force the old computation to occur. Yes.